File: day2/haskell/part1.hs
import System.Environment (getArgs)

main :: IO ()
main = do
    args <- getArgs
    let filename = if null args then "input.txt" else head args
    content <- readFile filename
    let ranges = map parseRange (takeWhile (/= "") (lines content))
        total = sum [n | (start, end) <- ranges, n <- [start..end], isInvalid n]
    print total

parseRange :: String -> (Int, Int)
parseRange line = (read start, read end)
  where
    (start, rest) = span (/= '-') line
    end = tail rest

isInvalid :: Int -> Bool
isInvalid n = even len && firstHalf == secondHalf
  where
    s = show n
    len = length s
    half = len `div` 2
    firstHalf = take half s
    secondHalf = drop half s

File: day2/go/part1.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

func isInvalid(n int) bool {
	s := strconv.Itoa(n)
	length := len(s)

	// Must have even length
	if length%2 != 0 {
		return false
	}

	half := length / 2
	firstHalf := s[:half]
	secondHalf := s[half:]

	return firstHalf == secondHalf
}

func main() {
	filename := "input.txt"
	if len(os.Args) > 1 {
		filename = os.Args[1]
	}

	file, _ := os.Open(filename)
	defer file.Close()

	scanner := bufio.NewScanner(file)
	scanner.Scan()
	line := scanner.Text()

	// Remove trailing comma if present
	line = strings.TrimSuffix(line, ",")

	ranges := strings.Split(line, ",")
	sum := 0

	for _, r := range ranges {
		parts := strings.Split(r, "-")
		start, _ := strconv.Atoi(parts[0])
		end, _ := strconv.Atoi(parts[1])

		for n := start; n <= end; n++ {
			if isInvalid(n) {
				sum = sum + n
			}
		}
	}

	fmt.Println(sum)
}
